File: js/patterns/fake_breakout.js
// js/patterns/fake_breakout.js

export class FakeBreakoutPattern {
    name = 'Ложный пробой уровня';
    confidence = 'medium';

    detect(candles) {
        // Увеличиваем окно до 50-100 свечей
        const windowSize = Math.min(100, Math.max(50, Math.floor(candles.length * 0.7)));
        // Расширяем окно подтверждения до 5-7 свечей
        const confirmationBars = Math.min(7, Math.max(5, Math.floor(candles.length * 0.05)));

        if (candles.length < windowSize + confirmationBars + 2) return null;

        // Берем предпоследнюю свечу для определения уровня
        const prevCandles = candles.slice(candles.length - windowSize - 1, candles.length - 1);
        const last = candles[candles.length - 1];

        const levelHigh = Math.max(...prevCandles.map(c => c.high));
        const levelLow = Math.min(...prevCandles.map(c => c.low));

        // Проверяем, сколько раз касались уровня
        const touchesHigh = prevCandles.filter(c => c.high >= levelHigh * 0.995).length;
        const touchesLow = prevCandles.filter(c => c.low <= levelLow * 1.005).length;

        // Берем свечи после текущей для проверки возврата
        const afterCurrent = candles.slice(candles.length - confirmationBars);

        // ============================================
        // ЛОЖНЫЙ ПРОБОЙ ВВЕРХ (SELL сигнал)
        // ============================================
        // НЕ требуем закрытия свечи за уровнем - достаточно касания или незначительного пробоя
        const fakeUp = last.high > levelHigh || last.high >= levelHigh * 0.998;
        const closeAbove = last.close > levelHigh;

        if (fakeUp && last.close > last.open) {
            // Проверяем, вернулась ли цена обратно в течение 5-7 свечей
            let returned = false;
            let returnDepth = 0;
            
            for (let i = 0; i < afterCurrent.length; i++) {
                const bar = afterCurrent[i];
                // Проверяем возврат ниже уровня (для ложного пробоя вверх)
                if (bar.close < levelHigh) {
                    returned = true;
                    // Вычисляем глубину возврата
                    returnDepth = (levelHigh - bar.close) / levelHigh;
                    break;
                }
            }

            if (returned) {
                // Проверяем объем на свече пробоя и на возврате
                const avgVolume = prevCandles.slice(-20).reduce((s, c) => s + (c.volume || 0), 0) / 20;
                const breakoutVolume = last.volume || 0;
                const returnVolume = afterCurrent.find(b => b.close < levelHigh)?.volume || 0;
                
                let confidence = this.confidence;
                if (breakoutVolume > avgVolume * 1.5 && returnVolume > avgVolume * 1.2) {
                    confidence = 'high';
                } else if (breakoutVolume > avgVolume * 1.2 || returnVolume > avgVolume * 1.2) {
                    confidence = 'medium';
                }

                // Уровень с 3+ касаниями дает более сильный сигнал
                if (touchesHigh >= 3) {
                    confidence = confidence === 'high' ? 'high' : 'medium';
                }

                const returnPercent = (returnDepth * 100).toFixed(1);
                return {
                    signal: 'SELL',
                    description: `Ложный пробой вверх: цена коснулась ${levelHigh.toFixed(2)} и вернулась на ${returnPercent}%`,
                    confidence: confidence
                };
            }
        }

        // ============================================
        // ЛОЖНЫЙ ПРОБОЙ ВНИЗ (BUY сигнал)
        // ============================================
        const fakeDown = last.low < levelLow || last.low <= levelLow * 1.002;
        const closeBelow = last.close < levelLow;

        if (fakeDown && last.close < last.open) {
            let returned = false;
            let returnDepth = 0;
            
            for (let i = 0; i < afterCurrent.length; i++) {
                const bar = afterCurrent[i];
                // Проверяем возврат выше уровня (для ложного пробоя вниз)
                if (bar.close > levelLow) {
                    returned = true;
                    returnDepth = (bar.close - levelLow) / levelLow;
                    break;
                }
            }

            if (returned) {
                const avgVolume = prevCandles.slice(-20).reduce((s, c) => s + (c.volume || 0), 0) / 20;
                const breakoutVolume = last.volume || 0;
                const returnVolume = afterCurrent.find(b => b.close > levelLow)?.volume || 0;
                
                let confidence = this.confidence;
                if (breakoutVolume > avgVolume * 1.5 && returnVolume > avgVolume * 1.2) {
                    confidence = 'high';
                } else if (breakoutVolume > avgVolume * 1.2 || returnVolume > avgVolume * 1.2) {
                    confidence = 'medium';
                }

                if (touchesLow >= 3) {
                    confidence = confidence === 'high' ? 'high' : 'medium';
                }

                const returnPercent = (returnDepth * 100).toFixed(1);
                return {
                    signal: 'BUY',
                    description: `Ложный пробой вниз: цена коснулась ${levelLow.toFixed(2)} и вернулась на ${returnPercent}%`,
                    confidence: confidence
                };
            }
        }

        return null;
    }
}

File: patterns/fake_breakout.py
# trade-signals-web/patterns/fake_breakout.py

import pandas as pd
from .base_pattern import BasePattern

class FakeBreakoutPattern(BasePattern):
    """Ложный пробой уровня"""
    
    def __init__(self):
        super().__init__("Ложный пробой уровня", confidence="medium")
    
    def get_min_candles(self) -> int:
        return 30
    
    def detect(self, df: pd.DataFrame) -> dict | None:
        window = 20
        confirmation_bars = 3
        
        if len(df) < window + confirmation_bars + 5:
            return None
        
        # Находим уровень
        level_high = df['high'].iloc[-window-1:-1].max()
        level_low = df['low'].iloc[-window-1:-1].min()
        
        candles_after_level = df.iloc[-window-1:].reset_index(drop=True)
        
        # Ложный пробой ВВЕРХ (SELL)
        for i in range(len(candles_after_level)):
            candle = candles_after_level.iloc[i]
            
            if candle['high'] > level_high and candle['close'] > level_high:
                if i + confirmation_bars >= len(candles_after_level):
                    break
                
                future_bars = candles_after_level.iloc[i+1:i+confirmation_bars+1]
                
                if future_bars['close'].min() <= level_high:
                    avg_volume = df['volume'].iloc[-window-1:-1].mean()
                    is_high_volume = candle['volume'] > avg_volume * 1.2
                    
                    return {
                        "signal": "SELL",
                        "description": f"Ложный пробой вверх: цена пробила {level_high:.2f} и вернулась обратно",
                        "confidence": "high" if is_high_volume else self.confidence
                    }
                else:
                    break
        
        # Ложный пробой ВНИЗ (BUY)
        for i in range(len(candles_after_level)):
            candle = candles_after_level.iloc[i]
            
            if candle['low'] < level_low and candle['close'] < level_low:
                if i + confirmation_bars >= len(candles_after_level):
                    break
                
                future_bars = candles_after_level.iloc[i+1:i+confirmation_bars+1]
                
                if future_bars['close'].max() >= level_low:
                    avg_volume = df['volume'].iloc[-window-1:-1].mean()
                    is_high_volume = candle['volume'] > avg_volume * 1.2
                    
                    return {
                        "signal": "BUY",
                        "description": f"Ложный пробой вниз: цена пробила {level_low:.2f} и вернулась обратно",
                        "confidence": "high" if is_high_volume else self.confidence
                    }
                else:
                    break
        
        return None
